src/Shared/Components/PrerenderComponentApplicationStore.cs

```diff
@@ -23,33 +30,103 @@ public PrerenderComponentApplicationStore(string existingState)
+        {
+            var state = JsonSerializer.Deserialize<Dictionary<string, byte[]>>(existingState);
+            if (state == null)
+            {
+                throw new ArgumentException("Could not deserialize state correctly", nameof(existingState));
+            }
+
+            var stateDictionary = new Dictionary<string, ReadOnlySequence<byte>>();
+            foreach (var (key, value) in state)
+            {
+                stateDictionary.Add(key, new ReadOnlySequence<byte>(value));
+            }
+
+            ExistingState = stateDictionary;
         }
 
 #nullable enable
-        public string? PersistedState { get; private set; }
+        public ReadOnlyMemory<char> PersistedState { get; private set; }
 #nullable disable
 
-        public Dictionary<string, byte[]> ExistingState { get; protected set; }
+        public Dictionary<string, ReadOnlySequence<byte>> ExistingState { get; protected set; }
 
-        public Task<IDictionary<string, byte[]>> GetPersistedStateAsync()
+        public Task<IDictionary<string, ReadOnlySequence<byte>>> GetPersistedStateAsync()
         {
-            return Task.FromResult((IDictionary<string, byte[]>)ExistingState);
+            return Task.FromResult((IDictionary<string, ReadOnlySequence<byte>>)ExistingState);
         }
 
-        protected virtual byte[] SerializeState(IReadOnlyDictionary<string, byte[]> state)
+        protected virtual PooledByteBufferWriter SerializeState(IReadOnlyDictionary<string, ReadOnlySequence<byte>> state)
         {
-            return JsonSerializer.SerializeToUtf8Bytes(state);
+            // System.Text.Json doesn't support serializing ReadonlySequence<byte> so we need to buffer
+            // the data with a memory pool here. We will change our serialization strategy in the future here
+            // so that we can avoid this step.
+            var buffer = new PooledByteBufferWriter();
+            var jsonWriter = new Utf8JsonWriter(buffer);
+            jsonWriter.WriteStartObject();
+            foreach (var (key, value) in state)
+            {
+                if (value.IsSingleSegment)
+                {
+                    jsonWriter.WriteBase64String(key, value.First.Span);
+                }
+                else
+                {
+                    WriteMultipleSegments(jsonWriter, key, value);
+                }
+                jsonWriter.Flush();
+            }
+
+            jsonWriter.WriteEndObject();
+            jsonWriter.Flush();
+            return buffer;
+
+            static void WriteMultipleSegments(Utf8JsonWriter jsonWriter, string key, ReadOnlySequence<byte> value)
+            {
+                byte[] unescapedArray = null;
+                var valueLenght = (int)value.Length;
+
+                Span<byte> valueSegment = value.Length <= 256 ?
+                    stackalloc byte[valueLenght] :
+                    (unescapedArray = ArrayPool<byte>.Shared.Rent(valueLenght)).AsSpan().Slice(0, valueLenght);
```

Comment:
Could we make this block more readable please?

Reply:
This is a well-known pattern to perform stackalloc with a fallback (the runtime code uses it all over the place). If you have a suggestion I'm open to it, however keep in mind that this can not be abstracted into a method since we are `stackalloc`ing